File: frontend/src/types/ws.ts
import type { QueryMetrics } from "./metrics";

export type WsClientMessage =
  | { type: "Subscribe"; payload: { topics: string[] } }
  | { type: "Unsubscribe"; payload: { topics: string[] } }
  | { type: "Ping" };

export type WsServerMessage =
  | { type: "MetricsUpdate"; payload: QueryMetrics }
  | { type: "K8sEvent"; payload: K8sEvent }
  | { type: "QueryProgress"; payload: QueryProgressEvent }
  | { type: "Pong" }
  | { type: "Error"; payload: { message: string } };

export interface K8sEvent {
  event_type: string;
  pod_name: string;
  message: string;
}

export interface QueryProgressEvent {
  query_id: string;
  progress_pct: number;
  rows_processed: number;
  stage: string;
}

File: crates/megafactory-sql-types/src/ws.rs
use serde::{Deserialize, Serialize};

/// Messages sent from the client to the WebSocket server.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", content = "payload")]
pub enum WsClientMessage {
    Subscribe { topics: Vec<String> },
    Unsubscribe { topics: Vec<String> },
    Ping,
}

/// Messages sent from the server to WebSocket clients.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", content = "payload")]
pub enum WsServerMessage {
    MetricsUpdate(crate::metrics::QueryMetrics),
    K8sEvent(K8sEvent),
    QueryProgress(QueryProgressEvent),
    Pong,
    Error { message: String },
}

/// A Kubernetes cluster event.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct K8sEvent {
    pub event_type: String,
    pub pod_name: String,
    pub message: String,
}

/// Progress update for a running query.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct QueryProgressEvent {
    pub query_id: String,
    pub progress_pct: f64,
    pub rows_processed: u64,
    pub stage: String,
}
